edpop_explorer/sparqlreader.py

```diff
@@ -50,90 +54,102 @@ def replace_fqu_with_prefixed_uris(inputstring: str) -> str:
 
-    def get_title(self) -> str:
-        return self.name
-
-    def show_record(self) -> str:
-        self.fetch()
-        field_strings = []
-        if self.link:
-            field_strings.append('URL: ' + self.link)
-        for field in self.fields:
-            fieldstring = replace_fqu_with_prefixed_uris(field)
-            field_strings.append(
-                '{}: {}'.format(fieldstring, self.fields[field])
-            )
-        return '\n'.join(field_strings)
+        self.fetched = True
 
-    def __repr__(self):
-        return self.get_title()
 
+class BibliographicalRDFRecord(RDFRecordMixin, BibliographicalRecord):
+    pass
 
-class SparqlReader(APIReader):
-    url: str = None
-    filter: str = None
-    wrapper: SPARQLWrapper
-    records: List[SparqlRecord]
-    name_predicate: str = None
 
-    def __init__(self):
-        self.wrapper = SPARQLWrapper(self.url)
-        self.wrapper.setReturnFormat(JSON)
+class SparqlReader(Reader):
+    endpoint: str
+    name_predicate: str
+    filter: Optional[str] = None
 
-    def prepare_query(self, query: str):
-        self.prepared_query = prepare_listing_query(
+    def transform_query(self, query: str):
+        return prepare_listing_query(
             name_predicate=self.name_predicate,
             filter=self.filter,
             query=query
         )
 
     def fetch(self):
         if not self.prepared_query:
-            raise APIException('First call prepare_query method')
-        self.wrapper.setQuery(self.prepared_query)
+            raise ReaderError('First call prepare_query method')
+        wrapper = SPARQLWrapper(self.endpoint)
+        wrapper.setReturnFormat(JSONFormat)
+        wrapper.setQuery(self.prepared_query)
         try:
-            response = self.wrapper.queryAndConvert()
+            response = wrapper.queryAndConvert()
         except SPARQLExceptions.QueryBadFormed as err:
-            raise APIException(
+            raise ReaderError(
                 'Malformed SPARQL query: {}'.format(err)
             )
+        assert isinstance(response, dict)
         results = response['results']['bindings']
         self.records = []
         self.number_of_results = len(results)
         for result in results:
-            record = SparqlRecord(
-                identifier=result['s']['value'],
-                sparql_endpoint=self.url,
-                link=result['s']['value'],
-                name=result['name']['value'],
-            )
-            self.records.append(record)
+            iri = result['s']['value']
+            name = result['name']['value']
+            self.records.append(self._create_lazy_record(iri, name))
         self.number_fetched = self.number_of_results
 
     def fetch_next(self):
         pass
+
+    @classmethod
+    @abstractmethod
+    def _convert_record(cls, graph: Graph, record: Record) -> None:
+        '''Convert data from an RDF graph to Fields in a Record. The 
+        Record is changed in-place.'''
```

Comment:
Hmm, I don't like this either, but this method is called from the `Record` itself when the full contents is being fetched, and I want to keep all the normalization logic inside the `Reader` classes. So while the readers with non-lazy records can simply create `Record` objects when fetching the data from the server, in case of readers with lazy records the `Record` objects are already there and have to be filled with data.

Subclassing `Record` would be an option as well, but that would not be consistent with the rest of the design (i.e. that there are no reader-specific `Record` subclasses and that all normalization is done by the `Reader`)...